solr/core/src/test/org/apache/solr/handler/admin/api/ReloadCoreAPITest.java

```diff
@@ -0,0 +1,155 @@
+  }
+
+  @Override
+  public Application configure() {
+    forceSet(TestProperties.CONTAINER_PORT, "0");
+    setUpMocks();
+    ResourceConfig config = new ResourceConfig();
+    config.register(ReloadCoreAPI.class);
+    config.register(CatchAllExceptionMapper.class);
+    config.register(SolrJacksonMapper.class);
+    config.register(NotFoundExceptionMapper.class);
+    config.register(
+        new AbstractBinder() {
+          @Override
+          protected void configure() {
+            bindFactory(new InjectionFactories.SingletonFactory<>(mockCoreContainer))
+                .to(CoreContainer.class)
+                .in(Singleton.class);
+          }
+        });
+    config.register(
+        new AbstractBinder() {
+          @Override
+          protected void configure() {
+            bindFactory(
+                    new Factory<SolrQueryRequest>() {
+                      @Override
+                      public SolrQueryRequest provide() {
+                        return new SolrQueryRequestBase(
+                            mockCoreContainer.getCore("demo"), new ModifiableSolrParams()) {};
+                      }
+
+                      @Override
+                      public void dispose(SolrQueryRequest instance) {}
+                    })
+                .to(SolrQueryRequest.class)
+                .in(RequestScoped.class);
+          }
+        });
+    config.register(
+        new AbstractBinder() {
+          @Override
+          protected void configure() {
+            bindFactory(
+                    new Factory<SolrQueryResponse>() {
+                      @Override
+                      public SolrQueryResponse provide() {
+                        return new SolrQueryResponse();
+                      }
+
+                      @Override
+                      public void dispose(SolrQueryResponse instance) {}
+                    })
+                .to(SolrQueryResponse.class)
+                .in(RequestScoped.class);
+          }
+        });
+    return config;
+  }
+
+  @Test
+  public void testValidReloadCoreAPIResponse() {
+    final Response response = target("/cores/demo/reload").request().post(Entity.json(""));
+    final String responseJsonStr = response.readEntity(String.class).toString();
+    assertEquals(200, response.getStatus());
+    assertEquals("{\"responseHeader\":{\"status\":0,\"QTime\":0}}", responseJsonStr);
+  }
+
+  @Test
+  public void testNonExistentCoreExceptionResponse() {
+    doThrow(
+            new SolrException(
+                SolrException.ErrorCode.BAD_REQUEST, "No such core: " + NON_EXISTENT_CORE))
+        .when(mockCoreContainer)
+        .reload("demo");
+    final Response response =
+        target("/cores/" + NON_EXISTENT_CORE + "/reload").request().post(Entity.json(""));
+    final String responseJsonStr = response.readEntity(String.class).toString();
+    assertEquals(400, response.getStatus());
+    assertTrue(responseJsonStr.contains("No such core"));
+  }
```

Comment:
@gerlowskija I added test cases to check the valid and error response but for some reason where I am throwing exception inside the API through mock, the test cases are failing. 
Upon further inspection, When API class throws error then It goes somehow to `CatchAllExceptionMapper.toResponse()` and there when it tries to get "solrQueryResponse" object containerRequestContext, it returns null.

Reply:
PR still in draft mode! When the test cases will be fixed. I will change it.

Reply:
Yeah, that's going to be pretty tough to mock out.  I think our best bet here is to test this at a slightly lower level (i.e. by calling ReloadCoreAPI.reloadCore directly and verifying the exception that it throws.  You've done similar stuff in other PRs, so I can use [your own code as an example](https://github.com/apache/solr/pull/1740/files#diff-ac75bc632270f1f0d082ebc4d82d4ffed7d162c2ec622aee7b38370fe960f2d2R86) 😛 )